[HWKINVENT-57] Move business logic to base impl inside API

metlos opened the pull request:
This is a big, big refactoring.
## What's wrong

There is a support for listening on delete events in the inventory. The problem with it is that only the entity being explicitly deleted is reported in the event. Any entities that are deleted along with it due to them being contained in the deleted entity (i.e. there being a `contains` relationship from the deleted entity to the contained entities, recursively) are NOT reported as such.

The only chance for the user of the inventory to figure out what will be deleted is to manually traverse the transitive closure over the contains edges from the entity to be deleted PRIOR to it being deleted. This is sub-optimal to say the least.
## How To Fix It

I've meant to do it for some time, but there was never a compelling reason to do it until now. The inventory as it exists at the moment implements large parts of the "business logic" (i.e. the semantic rules for different relationships, query composition, etc.) inside a concrete implementation that deals with a particular backend. But should we ever support another backend (like tinkerpop3 that we will have to support in order to use the future Titan 1.0), that logic will have to be duplicated in that implementation, too.

So the fix is to pull as much of the implementation of the API out of the tinkerpop impl into the API itself, currently into a package called "lazy" to suggest that the backend is queried lazily only when some actual results are needed and not during the individual "hops" through the inventory.

This will leave a rather small SPI interface that the real "backends" will have to implement.

Along with the change I also pulled out the actual test suite from the tinkerpop impl up into the API tests. A concrete backend implementation will then depend on the test-jar of the API and will inherit this abstract testsuite providing the implementation to test with otherwise standard test suite.

Once we have common business logic and behavior up in the API, we can easily add support for sending events during the delete without a fear that some other implementation will "forget" to do it because it will be the API itself that will be doing it.
## Progress
- [x] Pull out the impl
- [x] Make the testsuite pass
- [x] Rename "lazy" to something more meaningful like "ri" (reference impl) and strongly suggest using it instead of reimplementing the whole API.
- [x] Javadocs, javadocs and more javadocs.
- [x] Tests of the backend SPI itself.
- [x] Remove the "old" implementation from the Tinkerpop module.
- [x] Merge `ObservableInventory` into the main API
- [x] Fire additional events during the delete
- ~~[ ] Port the fix for HWKINVENT-255.~~ This needs to be revisited anyway once we start running on Titan in Hawkular.

jkremser commented on api/src/main/java/org/hawkular/inventory/lazy/Mutator.java:
```diff
@@ -0,0 +1,346 @@
+ * @since 0.0.6
+ */
+abstract class Mutator<BE, E extends Entity<Blueprint, Update>, Blueprint extends Entity.Blueprint,
+        Update extends AbstractElement.Update> extends Traversal<BE, E> {
```
`abstract class Mutator<BE, E extends Entity<Blueprint, Update>, Blueprint extends Entity.Blueprint, Update extends AbstractElement.Update> extends Traversal<BE, E>` hmmm :]

metlos replied:
Yeah, I know... javadoc would help here as would saner generics syntax ;)

metlos commented:
I need to rebase all the commits to rename the [HWKINVENT-45] under which I originally started fixing this to [HWKINVENT-57] which the fix ended up being implemented under.

metlos committed:
[HWKINVENT-57] First partial stab at lazy impl of inventory. This will most probably morph
into a single "possible" inventory implementation with different SPIs
backing it.

This will enable us to guarantee behavior across different backends.

metlos committed:
[HWKINVENT-57] Some more work on the lazy infrastructure.
EntityVisitor -> ElementVisitor + visitors for update and blueprint
types. This will help implementing the lazy interface because it will be
possible to use these to sanely separate the different impls for
different types of entities (instead of if-elsing on instanceof).

metlos committed:
[HWKINVENT-57] First stab at implementation of the lazy backend.
So far the "normal" impl is left intact and the lazy impl is included
in its own package for experimentation.

metlos committed:
[HWKINVENT-57] First set of tests for the lazy impl - so far only tenants are working..

metlos committed:
[HWKINVENT-57] Somehow, the persistent dir has moved itself to target.

metlos committed:
[HWKINVENT-57] Query generation fixes, environments support in lazy inventory.

metlos committed:
Cosmetic - checkstyle compliance, static imports.

metlos committed:
[HWKINVENT-57] Start of relationships support. This needs tests.

metlos committed:
[HWKINVENT-57] Associations implemented, all traversal methods done, tests enabled.

metlos committed:
[HWKINVENT-57] Random assortment of fixes to make the test suite pass.

metlos committed:
[HWKINVENT-57] Porting over the temporary fix for HWKINVENT-255.

metlos committed:
[HWKINVENT-57] Generate the test jars of the API.

metlos committed:
Oh, mighty checkstyle, I bow before thee.

metlos committed:
[HWKINVENT-57] Folding observable events into the main inventory interface

and implementing it in the reference lazy implementation.

metlos committed:
[HWKINVENT-57] Switch from "Lazy" to "Base" in the base impl of the
inventory.

metlos committed:
[HWKINVENT-57] Get rid of the Observable mixin.

This is now part of the core interface.

metlos committed:
[HWKINVENT-57] Remove the old Tinkerpop impl and replace it with the
new one based on the BaseInventory.

metlos committed:
Remove the workaround for HWKINVENTORY-255. The cause was different.

metlos committed:
Merge remote-tracking branch 'upstream/master' into bug/HWKINVENT-57

metlos committed:
Fix compilation error on Java8 prior to update 40.

metlos committed:
Merge remote-tracking branch 'upstream/master' into bug/HWKINVENT-57

metlos committed:
[HWKINVENT-57] Get rid of the inventory mixins. That was just a bad
idea.

Also, there is just a single "@Official" inventory instance to be
shared in CDI. REST builds on top of that to build a custom
"AutoTenant" implementation that suites the needs of the security
imposed by the REST API. AutoTenant is not needed anywhere else but
the REST api so it made little sense to have it right in the
component API.

metlos committed:
Trivial - @since update.

metlos committed:
[HWKINVENT-57] Javadocs and minor visibility and other cleanup.

metlos committed:
Trivial - avoid unchecked warnings.

metlos committed:
[HWKINVENT-57] Whoops, fix the compilation failure.

metlos committed:
Trivial - checkstyle compliance.

metlos committed:
[HWKINVENT-57] Removed unused backend methods.

metlos committed:
[HWKINVENT-57] Added a bunch of backend unit tests - query() is BROKEN!

metlos committed:
[HWKINVENT-57] Fixed query generation by properly tracking filter/path
boundaries.

metlos committed:
Merge remote-tracking branch 'upstream/master' into bug/HWKINVENT-57

jkremser commented on impl-tinkerpop-parent/impl/src/main/java/org/hawkular/inventory/impl/tinkerpop/TinkerpopBackend.java:
```diff
@@ -0,0 +1,728 @@
+    @Override
+    public Page<Element> query(Query query, Pager pager) {
+        HawkularPipeline<?, ? extends Element> q;
+        if (query.getFragments()[0].getFilter() instanceof RelationFilter) {
```
This looks more readable than the `JumpInOutFilter` filter. What is the `SwitchElementType` good for, it looks like the same thing as the old `JumpInOutFilter`, but here I can see it's not used anymore.

metlos replied:
It is the same thing as `JumpInOutFilter`. It is used in the base impl, afair. The backend just translates the Query to its native form - here a gremlin pipeline and executes it. There's still the `FilterApplicator` which mostly remained the same where you can see how it applies the `SwitchElementType` (which as you correctly found out is just a renamed `JumpInOutFilter`).

jkremser commented on api/src/main/java/org/hawkular/inventory/base/Mutator.java:
```diff
@@ -0,0 +1,431 @@
+        //report the relationship deletions first - it would be strange to report deletion of a relationship after
+        //reporting that an entity on one end of the relationship has been deleted
+        for (Relationship r : deletedRelationships) {
+            context.notify(r, deleted());
```
Isn't entity-based notification enough? Perhaps adding/removing an edge can be considered also as the entity update, I don't know.

metlos replied:
Well, I think people would be interested if two entities were no longer related. If you allow for relationship notifications, you need to be consistent and therefore emit them also during entity deletions and creations.

pilhuhn replied:
> Well, I think people would be interested if two entities were no 
> longer related. If you allow for relationship notifications, you need 
> to be consistent and therefore emit them also during entity deletions 
> and creations.

I agree. In the future we may have a UI that graphically shows 
components
of an app and potentially relations to other apps. So if you e.g. show
app as lb+3*as+db and have a link from db to other apps, you want
that graph get updated if one of the links from db to other apps goes
away.

jkremser commented on api/src/main/java/org/hawkular/inventory/base/Mutator.java:
```diff
@@ -0,0 +1,431 @@
+
+            context.backend.commit();
+        } catch (Throwable t) {
+            context.backend.rollback();
```
where does the transaction start? I recall some chat about it on IRC, but if it is the first call to the backend after previous commit/rollback, it's somewhat strange, because the read only calls don't necessarily need to open the transaction.

Is the graph smart enough to recognize `not-modifying op` and not open the transaction for them?

metlos replied:
Transactions in blueprints are thread-local and start with the first access to the database (read or write).
Which reminds me - we need to do `commit()`s even in the querying calls because otherwise we risk seeing stale data.
